tests/unit/backend/corpora/dataset_processing/test_process.py

```diff
@@ -178,6 +178,74 @@ def list_equal(list1, list2, cmp_func):
+    def test_extract_metadata_find_raw_layer(self, mock_read_h5ad):
+        # Setup anndata to be read
+        X_df = pandas.DataFrame(numpy.full((11, 3), 2), columns=list("ABC"), index=(str(i) for i in range(11)))
+        awesome_layer_df = pandas.DataFrame(
+            numpy.full((11, 3), 3), columns=list("ABC"), index=(str(i) for i in range(11))
+        )
+
+        obs = pandas.DataFrame(
+            numpy.hstack(
+                [
+                    numpy.array([["lung", "liver"][i] for i in numpy.random.choice([0, 1], size=(11))]).reshape(11, 1),
+                    numpy.array(
+                        [["UBERON:01", "UBERON:10"][i] for i in numpy.random.choice([0, 1], size=(11))]
+                    ).reshape(11, 1),
+                    numpy.array(
+                        [["10x", "smartseq", "cite-seq"][i] for i in numpy.random.choice([0, 1, 2], size=(11))]
+                    ).reshape(11, 1),
+                    numpy.array(
+                        [["EFO:001", "EFO:010", "EFO:011"][i] for i in numpy.random.choice([0, 1, 2], size=(11))]
+                    ).reshape(11, 1),
+                    numpy.random.choice(["healthy"], size=(11, 1)),
+                    numpy.random.choice(["MONDO:123"], size=(11, 1)),
+                    numpy.random.choice(["male", "female"], size=(11, 1)),
+                    numpy.array(
+                        [["solomon islander", "orcadian"][i] for i in numpy.random.choice([0, 1], size=(11))]
+                    ).reshape(11, 1),
+                    numpy.array(
+                        [["HANCESTRO:321", "HANCESTRO:456"][i] for i in numpy.random.choice([0, 1], size=(11))]
+                    ).reshape(11, 1),
+                    numpy.array(
+                        [["adult", "baby", "tween"][i] for i in numpy.random.choice([0, 1, 2], size=(11))]
+                    ).reshape(11, 1),
+                    numpy.array(
+                        [["HsapDv:0", "HsapDv:1", "HsapDv:2"][i] for i in numpy.random.choice([0, 1, 2], size=(11))]
+                    ).reshape(11, 1),
+                ]
+            ),
+            columns=[
+                "tissue",
+                "tissue_ontology_term_id",
+                "assay",
+                "assay_ontology_term_id",
+                "disease",
+                "disease_ontology_term_id",
+                "sex",
+                "ethnicity",
+                "ethnicity_ontology_term_id",
+                "development_stage",
+                "development_stage_ontology_term_id",
+            ],
+            index=(str(i) for i in range(11)),
+        )
+        uns = {
+            "title": "my test dataset",
+            "organism": "Homo sapiens",
+            "organism_ontology_term_id": "NCBITaxon:8505",
+            "layer_descriptions": {"my_awesome_wonky_layer": "raw"},
+        }
+        adata = anndata.AnnData(X=X_df, obs=obs, uns=uns, layers={"my_awesome_wonky_layer": awesome_layer_df})
```

Comment:
Do we expect `X_df` and `awesome_layer_df` have different nonzero genes per cell? It seems like they're both full, nonzero matrices.

Reply:
Yep -- lines 184 and 185 initialize the `X_df` and `awesome_layer_df` with 2s and 3s respectively. Which is why I assert below that the `mean_genes_per_cell` is 3 and not 2.

Reply:
Right, but `mean_genes_per_cell` is the mean count of nonzero genes per cell. Doesn't matter if it's 3, 2, or a million. I think the test is passing because the matrix has three columns.

Reply:
Ohhhhhhhhhhhhh

Reply:
Good catch -- will fix :P